executor/new_index_lookup_join.go

```diff
@@ -0,0 +1,534 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"unsafe"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/sessionctx/stmtctx"
+	"github.com/pingcap/tidb/terror"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/mvmap"
+	goctx "golang.org/x/net/context"
+)
+
+var _ = fmt.Sprint(nil)
+
+// NewIndexLookUpJoin employs one outer worker and N innerWorkers to execute concurrently.
+// It preserves the order of the outer table and support batch lookup.
+//
+// The execution flow is very similar to IndexLookUpReader:
+// 1. outerWorker read N outer rows, build a task and send it to result channel and inner worker channel.
+// 2. The innerWorker receives the task, builds key ranges from outer rows and fetch inner rows, builds inner row hash map.
+// 3. main thread receives the task, waits for inner worker finish handling the task.
+// 4，main thread join each outer row by look up the inner rows hash map in the task.
+type NewIndexLookUpJoin struct {
+	baseExecutor
+
+	resultCh   <-chan *lookUpJoinTask
+	cancelFunc goctx.CancelFunc
+	workerWg   *sync.WaitGroup
+
+	outerCtx outerCtx
+	innerCtx innerCtx
+
+	task             *lookUpJoinTask
+	joinResult       *chunk.Chunk
+	joinResultCursor int
+
+	resultGenerator joinResultGenerator
+}
+
+type outerCtx struct {
+	schema   *expression.Schema
+	rowTypes []*types.FieldType
+	keyCols  []int
+	filter   expression.CNFExprs
+}
+
+type innerCtx struct {
+	readerBuilder *dataReaderBuilder
+	rowTypes      []*types.FieldType
+	keyCols       []int
+	filter        expression.CNFExprs
+}
+
+type lookUpJoinTask struct {
+	outerResult *chunk.Chunk
+	outerMatch  []bool
+
+	innerResult       *chunk.List
+	encodedLookUpKeys *chunk.Chunk
+	lookupMap         *mvmap.MVMap
+	matchedInners     []chunk.Row
+
+	doneCh chan error
+	cursor int
+}
+
+type outerWorker struct {
+	outerCtx
+
+	ctx      context.Context
+	executor Executor
+
+	executorChk *chunk.Chunk
+
+	maxBatchSize int
+	batchSize    int
+
+	resultCh chan<- *lookUpJoinTask
+	innerCh  chan<- *lookUpJoinTask
+}
+
+type innerWorker struct {
+	innerCtx
+
+	taskCh      <-chan *lookUpJoinTask
+	outerCtx    outerCtx
+	ctx         context.Context
+	executorChk *chunk.Chunk
+}
+
+// Open implements the Executor interface.
+func (e *NewIndexLookUpJoin) Open(goCtx goctx.Context) error {
+	err := e.children[0].Open(goCtx)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	e.startWorkers(goCtx)
+	return nil
+}
+
+func (e *NewIndexLookUpJoin) startWorkers(goCtx goctx.Context) {
+	concurrency := e.ctx.GetSessionVars().IndexLookupConcurrency
+	resultCh := make(chan *lookUpJoinTask, concurrency)
```

Comment:
`e.resultCh = make(chan *lookUpJoinTask, concurrency)`

Reply:
Once the channel is assigned to a receive only type, it can not be converted to send only type.